server/src/cloud/auth.ts

```diff
@@ -0,0 +1,138 @@
+import {google} from "googleapis";
+import axios from "axios";
+import qs from "qs";
+import {publicApi} from "./util";
+
+const {GOOGLE_CLIENT_ID, GOOGLE_CLIENT_SECRET, GITHUB_CLIENT_ID, GITHUB_CLIENT_SECRET, AUTH_REDIRECT_URI, MICROSOFT_CLIENT_ID, MICROSOFT_CLIENT_SECRET} = process.env;
+
+const getGoogleOAuth2Client = () => {
+  const {OAuth2} = google.auth;
+  return new OAuth2(GOOGLE_CLIENT_ID, GOOGLE_CLIENT_SECRET, AUTH_REDIRECT_URI);
+};
+
+export interface UserInformation {
+  id: string;
+  name: string;
+  idToken?: string;
+  accessToken: string;
+  photoURL: string;
+}
+
+export const initializeAuthFunctions = (): void => {
+  if (GOOGLE_CLIENT_ID && GOOGLE_CLIENT_SECRET && AUTH_REDIRECT_URI) {
+    /**
+     * Function generates consent page URL where the code is retrieved
+     * github.com/autodidaktum/google-oauth2-parse-react/blob/master/deploy/cloud/main.js
+     * https://www.npmjs.com/package/googleapis#oauth2-client
+     */
+    publicApi<{state: string}, string>("GoogleSignIn", async ({state}) => {
+      const oauth2Client = getGoogleOAuth2Client();
+      return oauth2Client.generateAuthUrl({
+        access_type: "offline", // gets refresh_token
+        scope: ["openid", "profile"],
+        state,
+      });
+    });
+
+    /**
+     * 1) Function exchanges the code for access token.
+     * 2) Access token is used to retrieve user information.
+     */
+    publicApi<{code: string}, UserInformation>("GoogleVerifySignIn", async ({code}) => {
+      const oauth2Client = getGoogleOAuth2Client();
+      const {tokens} = await oauth2Client.getToken(code);
+      oauth2Client.setCredentials(tokens);
+      const oauth2 = google.oauth2({
+        auth: oauth2Client,
+        version: "v2",
+      });
+      const user = await oauth2.userinfo.get();
+      return {
+        id: user.data.id,
+        name: user.data.name,
+        idToken: tokens.id_token,
+        accessToken: tokens.access_token,
+        photoURL: user.data.picture,
+      };
+    });
+  }
+
+  if (GITHUB_CLIENT_ID && GITHUB_CLIENT_SECRET && AUTH_REDIRECT_URI) {
+    /**
+     * Function generates consent page URL where the code is retrieved
+     * https://docs.github.com/en/developers/apps/building-oauth-apps/authorizing-oauth-apps
+     */
+    publicApi<{state: string}, string>(
+      "GithubSignIn",
+      async ({state}) => `https://github.com/login/oauth/authorize?client_id=${GITHUB_CLIENT_ID}&scope=user&state=${state}&redirect_uri=${encodeURI(AUTH_REDIRECT_URI)}`
```

Comment:
What's the benefit? Readability?

Reply:
Yes. This is one benefit. And another benefit is that this will ensure that the URLs are valid.

Reply:
Validity has been tested. If escaping was a concern I would prefer your solution, too. In my opinion the one liner is far better readable and thus preferable.

Reply:
Normally you want to avoid building your own queries (e.g. sql queries, ... - in most cases for safety reasons). So i would stick to my solution and use the `URL` api for this purpose. And I think it's more readable if this long string is broken up into its separate parts (searchParams).

Reply:
Ok, lets do it the way @Dominik-Weinzierl proposed but skip the `encodeUri(...)` call, as it's not necessary with the URL class. So instead this code snippet should look like...

```
const url = new URL("/login/oauth/authorize", "https://github.com");
        url.searchParams.append("client_id", GITHUB_CLIENT_ID);
        url.searchParams.append("scope", "user");
        url.searchParams.append("state", state);
        url.searchParams.append("redirect_uri", AUTH_REDIRECT_URI);
        return url.toString();
```